Avoid saving extra register for IPInt JS->Wasm wrappers.

justinmichaud opened the pull request:
#### 7a1fe31ef5a7a4ebdbf52c296d76189fad22dd78
<pre>
Avoid saving extra register for IPInt JS-&gt;Wasm wrappers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274043">https://bugs.webkit.org/show_bug.cgi?id=274043</a>

Reviewed by Keith Miller and Yusuke Suzuki.

Adding this register to the pinned register set causes a bunch of places
to save it. We should just spill it inside the JS-&gt;WASM interpreted entry thunk.

* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::JSEntrypointInterpreterCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::makeInterpretedJSToWasmCallee):

Canonical link: <a href="https://commits.webkit.org/279435@main">https://commits.webkit.org/279435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d70c27517cf6539d2c0065fa80d2b2ec831f6c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5999 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4219 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4007 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2768 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55591 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24496 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3550 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2375 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46853 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3722 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58368 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53008 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3727 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50766 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29857 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/46424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11660 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30786 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65313 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29629 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12449 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -579,7 +579,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, useWasmIPIntEpilogueOSR, true, Normal, "Allow IPInt to tier up during function epilogues"_s) \
     v(Bool, wasmIPIntTiersUpToBBQ, true, Normal, "Allow IPInt to tier up to BBQ"_s) \
     v(Bool, wasmIPIntTiersUpToOMG, true, Normal, "Allow IPInt to tier up to OMG"_s) \
-    v(Bool, useIPIntWrappers, false, Normal, "Allow IPInt to replace JIT wasm wrappers"_s) \
```
This is just for the EWS run

justinmichaud replied:
Although after Max's alignment change, maybe we can turn this on by default without it breaking stuff

Achierius commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -604,17 +604,29 @@ op(js_to_wasm_wrapper_entry, macro ()
+    # any callee saves used here would also have to be saved by the JS->WASM thunk,
+    # since exceptions cannot tell if an interpreted callee had a replacementCallee installed
+    # when the stack frame was created.
+    # We should just restore extra callee-saves before calling.
```
What do you mean 'extra callee saves' -- above you say 'any callee saves', is there a subset of them which we wouldn't need to do that restore?

justinmichaud replied:
Updated wording

Achierius commented on Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
```diff
@@ -123,15 +123,8 @@ bool LLIntPlan::makeInterpretedJSToWasmCallee(unsigned functionIndex)
-    }
 #if CPU(ARM64) || CPU(ARMv7)
-    const size_t JSEntrypointInterpreterCalleeSaveSpaceStackAligned = WTF::roundUpToMultipleOf(stackAlignmentBytes(), 4 * sizeof(CPURegister));
+    const size_t JSEntrypointInterpreterCalleeSaveSpaceStackAligned = WTF::roundUpToMultipleOf(stackAlignmentBytes(), 6 * sizeof(CPURegister));
```
That might explain an issues I ran into. What do you think about adding a comment pointing here from up where the spilling happens?

kmiller68 replied:
This feels impossible to find. Can we have a list of registers saved by JSToWasm in RegisterSet and reference that here?

justinmichaud replied:
I just did sp-based indexing instead, so now there is only one place with this

Achierius reviewed (approved):
r=me
albeit only informally

kmiller68 commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -421,7 +421,7 @@ RegisterAtOffsetList* JSEntrypointInterpreterCallee::calleeSaveRegistersImpl()
 #if CPU(X86_64)
 #elif CPU(ARM64) || CPU(RISCV64)
-        ASSERT(registers.numberOfSetRegisters() == 4);
+        ASSERT(registers.numberOfSetRegisters() == 3);
```
Why is this 3 but JSEntrypointInterpreterCalleeSaveSpaceStackAligned is 6?

justinmichaud replied:
This is the set of callee-save registers that are allowed to be trashed by the thunk. Every other callee save needs to be restored before calling into JS, in case we throw an exception. I'll find better names

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -604,17 +604,29 @@ op(js_to_wasm_wrapper_entry, macro ()
 
     const AccumulatorTag = invalidGPR # Only used for JSVALUE32_64
```
It's weird that the option for this is called IPInt but this code is in the LLInt implementation.

justinmichaud replied:
Changed

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +653,57 @@ end
         if ARM64 or ARM64E
-            storepairq boundsCheckingSize, metadataTable, -16[cfr]
-            storepairq wasmInstance, memoryBase, -32[cfr]
+            storepairq memoryBase, boundsCheckingSize, -16[cfr]
```
It's weird these are in decimal and the numbers below are in hex.

Anyway, can we do `-X * SlotSize[cfr]` for all these?

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +653,57 @@ end
+
 if ARM64 or ARM64E or ARMv7
-    const JSEntrypointInterpreterCalleeSaveSpaceStackAligned = (4 * SlotSize + StackAlignment - 1) & ~StackAlignmentMask
+    const JSEntrypointInterpreterSpillSpaceStackAligned = (6 * SlotSize + StackAlignment - 1) & ~StackAlignmentMask
```
Shouldn't this be 5 (maybe 4)? boundsCheckingSize, metadataTable, wasmInstance, MPC, ws0 (ws0 is caller save though so I don't know why it's getting saved). It would also be nice if this was with the register naming you / spill slot assignment you do above.

justinmichaud replied:
I should maybe leave a comment, ws0 was just padding. I guess we could just do a single store here

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +653,57 @@ end
-            loadpairq -16[cfr], boundsCheckingSize, metadataTable
-            loadpairq -32[cfr], wasmInstance, memoryBase
+            loadpairq -16[cfr], memoryBase, boundsCheckingSize
+            loadpairq -32[cfr], ws0, wasmInstance
```
Why did this order change? Isn't the order determined by the register number?

justinmichaud replied:
The arm spec says operand order, unless offline asm is doing something scary. That is important too, because this needs to match whatever the JSToWasm stub does. In either case, the order should be the same, just moved up by one slot. I would not be surprised if I got the order wrong though

justinmichaud replied:
Anyway, EWS caught this when I wrote the original patch, so it is probably right

kmiller68 replied:
I'm not exactly sure what the ARM spec says or that it really matters (we already differ from the ARM C ABI in a variety of ways). What mostly matters that this matches what genericUnwind thinks the spill order is, which is determined by RegisterAtOffsetList. RegisterAtOffsetList fills the values in the register enum order so that's what LLInt needs to match in order to work correctly with genericUnwind.

justinmichaud replied:
Ah, I misunderstood. Yes, exactly.

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +653,57 @@ end
-            storepairq boundsCheckingSize, metadataTable, -16[cfr]
-            storepairq wasmInstance, memoryBase, -32[cfr]
+            storepairq memoryBase, boundsCheckingSize, -16[cfr]
+            storepairq ws0, wasmInstance, -32[cfr]
```
Isn't ws0 caller save?

justinmichaud replied:
Yes, we don't need to save it. I can just do a single store

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -605,16 +605,29 @@ op(js_to_wasm_wrapper_entry, macro ()
     const Accumulator = csr2
     const Scratch = ws1
+
+    # Any callee saves clobbered here would also have to be saved by the JS->WASM thunk,
+    # since the exception unwinder cannot tell if an interpreted callee had a replacementCallee installed
+    # when the stack frame was created.
+    # Instead of making every thunk save more callee-saves, we should just restore these extra
+    # clobbered callee-saves here before calling JS.
```
Can we file a bug to fix this once we have GlobalGC and can make Wasm::Callee GC'd?

justinmichaud replied:
Do you mean that we would replace the interpreted callee directly, instead of via m_replacementCallee? Or that we would mark weather the frame had a replacementCallee when it was created?

kmiller68 replied:
The former, the upside would be twofold. 1) We wouldn't need to match the spills between the two JSToWasm callees. 2) We could actually free lower tier Wasm code that's been replaced as the GC'd Callee on the stack would keep the JIT code alive while it is still reachable.

justinmichaud replied:
Nice, sure

justinmichaud replied:
https://bugs.webkit.org/show_bug.cgi?id=274768

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +653,57 @@ end
+            storep wasmInstance, -0x18[cfr]
             # This gets restored before call too, since it doesn't get saved/restored by the wasmToJS thunk
             storep Accumulator, AccumulatorCSRSlot[cfr]
+            storep MPC, MPCCSRSlot[cfr]
```
Can we flip these so they're in the order they're actually spilled?

kmiller68 reviewed (commented):
r=me.

kmiller68 commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -640,54 +656,57 @@ end
-            loadpairq -16[cfr], boundsCheckingSize, metadataTable
-            loadpairq -32[cfr], wasmInstance, memoryBase
+            loadpairq -2 * SlotSize[cfr], memoryBase, boundsCheckingSize
+            loadpairq -4 * SlotSize[cfr], ws0, wasmInstance
```
Nit: It's a little weird this does a loadpair but we do a single store above.

MenloDorian replied:
Why is this weird?  This seems appropriate for ARM64.  What am I missing?

MenloDorian replied:
Oh, the extra load into ws0.  That looks unneeded.

justinmichaud replied:
Nice catch, I missed this one. Thanks!

kmiller68 reviewed (approved):
r=me with nit.

Constellation reviewed (approved):
r=me

justinmichaud committed:
Avoid saving extra register for IPInt JS->Wasm wrappers.
https://bugs.webkit.org/show_bug.cgi?id=274043

Reviewed by Keith Miller and Yusuke Suzuki.

Adding this register to the pinned register set causes a bunch of places
to save it. We should just spill it inside the JS->WASM interpreted entry thunk.

* Source/JavaScriptCore/jit/RegisterSet.cpp:
(JSC::RegisterSetBuilder::wasmPinnedRegisters):
* Source/JavaScriptCore/llint/WebAssembly.asm:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::JSEntrypointInterpreterCallee::calleeSaveRegistersImpl):
* Source/JavaScriptCore/wasm/WasmLLIntPlan.cpp:
(JSC::Wasm::LLIntPlan::makeInterpretedJSToWasmCallee):

Canonical link: https://commits.webkit.org/279435@main
